src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/Models/ActiveDirectoryClient.cs

```diff
@@ -0,0 +1,818 @@
+        {
+            PSADServicePrincipal servicePrincipal = null;
+            try
+            {
+                var odataQuery = new Rest.Azure.OData.ODataQuery<ServicePrincipal>(s => s.ServicePrincipalNames.Contains(spn));
+                servicePrincipal = GraphClient.ServicePrincipals.List(odataQuery.ToString()).FirstOrDefault()?.ToPSADServicePrincipal();
+            }
+            catch { /* The service principal does not exist, ignore the exception. */ }
+
+            return servicePrincipal;
+        }
+
+        public IEnumerable<PSADServicePrincipal> FilterServicePrincipals(Rest.Azure.OData.ODataQuery<ServicePrincipal> odataQuery, ulong first = ulong.MaxValue, ulong skip = 0)
+        {
+            return new GenericPageEnumerable<ServicePrincipal>(
+                delegate ()
+                {
+                    return GraphClient.ServicePrincipals.List(odataQuery);
+                }, GraphClient.ServicePrincipals.ListNext, first, skip).Select(s => s.ToPSADServicePrincipal());
+        }
+
+        public IEnumerable<PSADServicePrincipal> FilterServicePrincipals(ADObjectFilterOptions options, ulong first = ulong.MaxValue, ulong skip = 0)
+        {
+            List<PSADServicePrincipal> servicePrincipals = new List<PSADServicePrincipal>();
+            ServicePrincipal servicePrincipal = null;
+
+            if (!string.IsNullOrEmpty(options.Id))
+            {
+                try
+                {
+                    servicePrincipal = GraphClient.ServicePrincipals.Get(options.Id);
+                }
+                catch {  /* The user does not exist, ignore the exception. */ }
+
+                if (servicePrincipal != null)
+                {
+                    servicePrincipals.Add(servicePrincipal.ToPSADServicePrincipal());
+                }
+            }
+            else if (!string.IsNullOrEmpty(options.SPN))
+            {
+                try
+                {
+                    var odataQuery = new Rest.Azure.OData.ODataQuery<ServicePrincipal>(s => s.ServicePrincipalNames.Contains(options.SPN));
+                    servicePrincipal = GraphClient.ServicePrincipals.List(odataQuery.ToString()).FirstOrDefault();
+                }
+                catch {  /* The user does not exist, ignore the exception. */ }
+
+                if (servicePrincipal != null)
+                {
+                    servicePrincipals.Add(servicePrincipal.ToPSADServicePrincipal());
+                }
+            }
+            else
+            {
+                Rest.Azure.OData.ODataQuery<ServicePrincipal> odataQuery = null;
+                if (!string.IsNullOrEmpty(options.SearchString) && options.SearchString.EndsWith("*"))
+                {
+                    options.SearchString = options.SearchString.TrimEnd('*');
+                    odataQuery = new Rest.Azure.OData.ODataQuery<ServicePrincipal>(s => s.DisplayName.StartsWith(options.SearchString));
```

Comment:
Does DisplayName always have a value?  If not, we should guard nulls

Reply:
I think it always has a value, but I added a null-check just in case.